allennlp/models/encoder_decoders/wikitables_semantic_parser.py

```diff
@@ -269,6 +339,57 @@ def forward(self,  # type: ignore
             # TODO(matt): compute accuracy here.
             return outputs
 
+    def _compute_average_with_encoder(self, embed_tensor: Variable,
+                                      mask: Variable) -> Variable:
+        # computes the average along dimension 1 of the embed_tensor
+        dims = embed_tensor.size()
+        table_mask = mask.view(-1, mask.size(2))
+        averaged_tensor = self._entity_encoder(embed_tensor.view(-1, dims[2], dims[3]),
+                                               table_mask)
+        # (batch_size, num_entities, embedding_dim)
+        return averaged_tensor.view(dims[0], dims[1], dims[3])
+
+    def _get_neighbor_indexes(self, world: List[WikiTablesWorld],
+                                    max_num_entities: int) -> Variable:
+        # returns each entity's neighbor indexes
+        batches = []
+        for w in world:
+            # each batch has it's own world aka table
+            entities = w.table_graph.entities
+            entity2idx = {e: i for i, e in enumerate(entities)}
+            neighbors = w.table_graph.neighbors
+            neighbor_indexes = []
+            for e in entities:
+                # get indexes of entity e's neighbors
+                curr_neighbors = [entity2idx[cn] for cn in neighbors[e]]
+                # pad to length of max number of entities in a batch
+                padded = pad_sequence_to_length(curr_neighbors, max_num_entities)
+                neighbor_indexes.append(padded)
+            # pad the list with empty lists to hit length max num entities
+            neighbor_indexes = pad_sequence_to_length(neighbor_indexes,
+                                                      max_num_entities,
+                                                      lambda:[0]*max_num_entities)
+            batches.append(neighbor_indexes)
+        # (batch_size, num_entities, num_entities)
+        return Variable(torch.LongTensor(batches))
```

Comment:
You can't actually create variables this way, because they might end up on the wrong device..  You need to do the `.data.new()` stuff seen here:
https://github.com/allenai/allennlp/blob/f9db6bc46e43930577e1cd6dc0fa8eb444a73681/allennlp/models/encoder_decoders/wikitables_semantic_parser.py#L1017

Reply:
Hmm, that's a good point.  Two options: (1) pass in a tensor here that you can copy from, and just say in the docstring that the only reason we have it is to copy it, or (2) return the padded array from this method, and actually make a tensor out of it in `forward`, using one of the tensors that's there.  I'd probably pick 1, just so the method is self-contained, but I could go either way.